### PR TITLE
Derive MRRA via gradient descent

### LINKED ISSUE #1990: Control the number of cores used in production
On the grid it is important to know in advance the amount of cores used in each job. In particular we found it is significantly more efficient to submit jobs which use just one core (due to memory assignment). 

In the current legacy framework there is a "sequential" mode which makes sure that CORSIKA and sim_telarray do not run simultaneously, but rather wait for each other to "pause". The same happens with multiple instances of sim_telarray. 

The same feature or even more control should be included in simtools. Not sure if this feature is already requested (I think it does not exist). Apologies if I missed it.

### PR DESCRIPTION
**Pull request by EshitaJoshi**
Use per-mirror d80 measurements to derive```mirror_reflection_random_angle``` via gradient descent.

1. Gradients are computed via central finite differences for each RNDA parameter.
    - $\sigma_1$ and $\sigma_2$ are scaled in log space because the values are very small $\sim 10^{-3}$ deg
    - frac is scaled linearly as it ranges from 0 to 1

2. Gradients are clipped to avoid instability due to very large updates.
    - Update steps are bounded (MAX_LOG_STEP, MAX_FRAC_STEP) and all parameters are clipped to their schema-defined bounds.

3. Adaptive learning rate is used
    - If the objective improves, the learning rate is increased slightly.
    - If not, the step is rejected and the learning rate is reduced.

4. Optimization stops when either:
    - The absolute percentage difference falls below the given threshold, or
    - The learning rate becomes negligible (1e-12), or
    - The maximum number of iterations is reached (100)

Usually takes around ~10 iterations per mirror to converge.

To run use:

```
simtools-derive-mirror-rnda --test --site North --telescope LSTN-01 --model_version 7.0.0 --data tests/resources/MLTdata-preproduction.ecsv --parameter_version 1.0.0 --d80_hist --cleanup
```

The per-mirror measured and simulated d80 distributions look like:

d80_distributions

The final averaged value of the parameter is exported like this in the telescope sub-folder: [mirror_reflection_random_angle-1.0.0.json](https://github.com/user-attachments/files/24668844/mirror_reflection_random_angle-1.0.0.json)

and the per-mirror optimization is exported like this:
[per_mirror_rnda.json](https://github.com/user-attachments/files/24668852/per_mirror_rnda.json)

The algorithm is parallelised, using the ```--n_workers``` flag or ```os.cpu_count()``` by default.

The parallelisation is generalised and moved to simtools.job_execution. Closes #1990.

### CODE REVIEW

File: src/simtools/applications/derive_mirror_rnda.py [hunk 1]
```diff
@@ -221,9 +174,22 @@ def main():
...
+
+    # Only write JSON file if optimization converged below threshold
+    threshold = app_context.args.get("threshold")
+    if panel_psf.final_rmsd <= threshold:
+        panel_psf.write_optimization_data()
+    else:
+        print(
+            f"\nSkipping parameter export: "
+            f"RMSD {panel_psf.final_rmsd:.6f} > threshold {threshold:.6f}"
+        )
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/applications/derive_mirror_rnda.py]: Fixed

File: src/simtools/ray_tracing/psf_parameter_optimisation.py [hunk 2]
```diff
@@ -581,19 +600,23 @@ def run_gradient_descent(self, rmsd_threshold, learning_rate, max_iterations=200
...
 
-            # Step was accepted - update state
+            # Step was accepted - update state and potentially increase learning rate
             current_params = step_result.params
             current_metric = step_result.metric
             current_psf_diameter = step_result.psf_diameter
             current_lr = step_result.learning_rate
 
+            # Increase learning rate after successful step to accelerate convergence
+            current_lr = self._increase_learning_rate(current_lr)
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/psf_parameter_optimisation.py]: Fixed, dont increase LR if it was reduced in the last step

File: tests/integration_tests/config/derive_mirror_rnda_psf_measurement.yml [hunk 3]
```diff
@@ -2,26 +2,19 @@
...
-    rtol_psf_containment: 0.1
+    data: tests/resources/PSFcurve_data_v2.ecsv
+    mirror_list: tests/resources/mirror_list_CTA-N-LST1_v2019-03-31_rotated.ecsv
     site: North
-    telescope: MSTx-NectarCam
+    telescope: LSTN-01
     test: true
-    use_random_focal_length: true
+    use_random_focal_length: false
   integration_tests:
```
Comment (GernotMaier) [inline tests/integration_tests/config/derive_mirror_rnda_psf_measurement.yml]: Wondering if we want to test that the output values (three model parameter values) is consistent with expectations? This would require to put an example json file into tests/resources/model_parameters and add a reference_output_file point like in tests/integration_tests/config/db_get_parameter_from_db_telescope_parameter_version.yml (don't ask about the formatting in this file)

File: src/simtools/applications/derive_mirror_rnda.py [hunk 4]
```diff
@@ -88,45 +58,46 @@
...
-            --telescope LSTS-design \\
-            --model_version 6.0.0 \\
-            --containment_fraction 0.8 \\
-            --mirror_list ./tests/resources/mirror_list_CTA-N-LST1_v2019-03-31_rotated.ecsv
-            --rnda 0.003 \\
-            --psf_measurement_containment_mean 1.4 \\
+            --site North \\
+            --telescope LSTN-01 \\
+            --model_version 6.0.2 \\
+            --mirror_list tests/resources/mirror_list_CTA-N-LST1_v2019-03-31_rotated.ecsv \\
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/applications/derive_mirror_rnda.py]: The mirror list provided here is to get the focal lengths of the mirrors (alternatively use ```--use_random_focal_length```). Running with ```--test``` we use the first 10 mirrors, but without it we use all mirrors. Each mirror is simulated separately but then the photons are all appended in one place, so we get one combined PSF image which we use for the optimisation. It's not an average, it's generating one PSF image from the total photon distribution.

**Comment by EshitaJoshi** [comment]
@GernotMaier could you look at this PR again please?

File: src/simtools/applications/derive_mirror_rnda.py [hunk 5]
```diff
@@ -1,227 +1,174 @@
...
+
+This application derives the value of the simulation model parameter
+*mirror_reflection_random_angle* using measurements of the d80 (spot size)
+and focal length of individual mirror panels.
+
+The optimization uses percentage difference as the metric::
+
+    pct_diff = 100 * (simulated_d80 - measured_d80) / measured_d80
+
+Each mirror is optimized individually, and the final RNDA is the average of all
```
Comment (GernotMaier) [inline src/simtools/applications/derive_mirror_rnda.py]: Although clear, RNDA is not defined?

File: src/simtools/ray_tracing/ray_tracing.py [hunk 6]
```diff
@@ -468,6 +468,34 @@ def export_results(self):
...
+        float
+            d80 in millimeters.
+        """
+        if self._results is None:
+            raise RuntimeError("No results available; run analyze() first")
+
+        d80 = self._results["d80_cm"][row_index]
+
+        # d80 is typically an astropy Quantity. Fall back gracefully if not.
+        if hasattr(d80, "to_value"):
```
Comment (GernotMaier) [inline src/simtools/ray_tracing/ray_tracing.py]: Trying to understand: you are trying to guess if this is a astropy quantity or a numpy array? If yes, wouldn't it be clearer to use `isinstance()` in both cases?

File: src/simtools/simtel/simulator_ray_tracing.py [hunk 7]
```diff
@@ -218,20 +222,19 @@ def _check_run_result(self, run_number=None):  # pylint: disable=unused-argument
...
-            file.write("Trigger 1 of 0\n")
+    def _write_out_single_pixel_camera_files(self):
+        """Write out per-label single pixel camera + funnel files.
 
-        # need to also write out the funnel_perfect.dat file
-        with self.telescope_model.config_file_directory.joinpath("funnel_perfect.dat").open(
-            "w"
-        ) as file:
+        These files are referenced by sim_telarray and must not be shared across
+        parallel worker processes (otherwise they can be truncated mid-read).
```
Comment (GernotMaier) [inline src/simtools/simtel/simulator_ray_tracing.py]: How do you make sure that they are not shared across work processes? The information is this is taken care of (e.g., by separate directories or file names) or not.

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 8]
```diff
@@ -74,220 +145,690 @@ def _define_telescope_model(self, label):
...
+        self.telescope_model.overwrite_model_parameter(
+            "mirror_reflection_random_angle", rnda_values
+        )
+
+        mirror_sim_index = int(mirror_idx)
+        mirror_number = mirror_sim_index + 1  # user-facing label
+
+        # Create ray tracing for single mirror.
+        # Use a per-mirror label so parallel runs do not collide on output filenames.
+        # IMPORTANT: RayTracing and SimulatorRayTracing derive filenames from
```
Comment (GernotMaier) [inline src/simtools/ray_tracing/mirror_panel_psf.py]: Is this something we should change in RayTracing and SimulatorRayTracing in a future pull request?

File: tests/resources/198mir_190925.ecsv
Comment (GernotMaier) [inline tests/resources/198mir_190925.ecsv]: Could you please add additional metadata describing this file?

- what measurement?
- for which telescope?
- author, origin?
- anything else which might be useful.

(and I am completely aware that this is missing for a couple of files in tests/resources. This is a problem (see also issue #1600).

Comment (GernotMaier) [inline tests/resources/198mir_190925.ecsv]: Is this file conceptionally different to tests/resources/MLTdata-preproduction.ecsv ?

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 9]
```diff
@@ -74,220 +151,691 @@ def _define_telescope_model(self, label):
...
+        measured_d80_mm: float,
+        focal_length_m: float,
+        current_rnda,
+        settings: RndaGradientDescentSettings,
+    ):
+        """Run a 3-parameter RNDA gradient descent (sigma1, fraction2, sigma2).
+
+        Returns best_rnda, best_sim_d80, best_pct_diff.
+        """
+        # pylint: disable=too-many-locals
```
Comment (GernotMaier) [inline src/simtools/ray_tracing/mirror_panel_psf.py]: too many pylint statements.

**Review by GernotMaier — CHANGES_REQUESTED** [review]
Thanks @EshitaJoshi , lots of very good changes! As discussed already in person, we need to work a bit on the maintainability of the new code. Shorter functions are easier. Plotting should be in a separate module. Think about how the result will be used (the log is good to read, but hard to work for with the computer). Consider if we need that many database calls.

File: tests/resources/198mir_190925.ecsv
Comment (EshitaJoshi) [reply tests/resources/198mir_190925.ecsv]: tests/resources/MLTdata-preproduction.ecsv doesn't have the focal lengths per mirror, which are used in the simulations

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 10]
```diff
@@ -74,220 +151,691 @@ def _define_telescope_model(self, label):
...
-        """Get optimization starting value from command line or previous model."""
-        if self.args_dict["rnda"] != 0:
-            rnda_start = self.args_dict["rnda"]
-        else:
-            rnda_start = self.telescope_model.get_parameter_value("mirror_reflection_random_angle")[
-                0
-            ]
+        return lines
+
+    def write_results_log(self):
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/mirror_panel_psf.py]: This log file was just meant to be a human readable summary of the run. There is also a .json file for the model parameter export and another .json file for the per-mirror optimisation data. Should I remove this log file?

File: src/simtools/ray_tracing/mirror_panel_psf.py
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/mirror_panel_psf.py]: Thanks, fixed

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 11]
```diff
@@ -1,26 +1,81 @@
...
+def _worker_init(label, args_dict, db_config):
+    """Initialize per-process MirrorPanelPSF instance.
+
+    Important: DB configuration is stored in ``simtools.settings.config`` at runtime.
+    With the multiprocessing "spawn" start method, workers do not inherit that state,
+    so we must call ``config.load`` again in each worker.
+    """
+    global _WORKER_INSTANCE  # pylint: disable=global-statement
+    settings_config.load(args=args_dict, db_config=db_config)
+    _WORKER_INSTANCE = MirrorPanelPSF(label=label, args_dict=args_dict)
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/mirror_panel_psf.py]: Now all the workers share a parent instance of MirrorPanelPSF

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 12]
```diff
@@ -74,220 +151,691 @@ def _define_telescope_model(self, label):
...
+            signed_pct_diff=signed_pct,
+            objective=obj,
+        )
 
-    def _get_psf_containment(self):
-        """Read measured single-mirror point-spread function from file and return mean and sigma."""
-        # If this is a test, read just the first few lines since we only simulate those mirrors
-        data_end = (
-            self.args_dict["number_of_mirrors_to_test"] + 1 if self.args_dict["test"] else None
+    def _finite_difference_objective_gradient(
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/mirror_panel_psf.py]: I was not able to reproduce the speed/accuracy with scipy.
d80_distributions_scipy

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 13]
```diff
@@ -1,294 +1,412 @@
...
-        ray.analyze(force=True)
-        if save_figures:
-            ray.plot("d80_cm", save=True, d80=self.args_dict["psf_measurement_containment_mean"])
-
-        return (
-            ray.get_mean("d80_cm").to(u.cm).value,
-            ray.get_std_dev("d80_cm").to(u.cm).value,
-        )
+        param_min, param_max = param_bounds
+        is_log_param = param_idx != 1
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/mirror_panel_psf.py]: fixed

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 14]
```diff
@@ -1,294 +1,412 @@
...
-                "mirror_reflection_random_angle_sigma2",
-                f"containment_radius_D{containment_fraction_percent}",
-                f"containment_radius_sigma_D{containment_fraction_percent}",
-            ),
+        n_mirrors = len(self.measured_data)
+        if self.args_dict.get("test"):
+            n_mirrors = min(n_mirrors, self.args_dict.get("number_of_mirrors_to_test", 1))
+
+        n_workers = int(self.args_dict.get("n_workers") or os.cpu_count())
+        parent = MirrorPanelPSF(self.label, dict(self.args_dict, parallel=False))
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/mirror_panel_psf.py]: fixed

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 15]
```diff
@@ -1,294 +1,394 @@
-"""Mirror panel PSF calculation."""
+"""Mirror panel PSF calculation with per-mirror d80 optimization."""
```
Comment (GernotMaier) [inline src/simtools/ray_tracing/mirror_panel_psf.py]: Is there actually a reason why we hardwire d80 and don't have the containment fraction a configurable parameter (with 80% as default)?

**Review by GernotMaier — APPROVED** [review]
Very nice! I can now understand what is happening. A couple of minor comments, nothing big so I approve.

For the plotting: is there still functionality to plot? The important point is to have all plotting related in modules of visualize, but we still want that functionality. Did I miss it?

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 16]
```diff
@@ -1,294 +1,394 @@
...
+
+    def _simulate_single_mirror_d80(self, mirror_idx, rnda_values):
+        """
+        Simulate a single mirror and return its d80.
 
-        return tel_model, site_model
+        Parameters
+        ----------
+        mirror_idx : int
+            Zero-based index of the mirror.
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/mirror_panel_psf.py]: Thinking on it again, i think its best to have it all be zero-based

File: src/simtools/ray_tracing/mirror_panel_psf.py [hunk 17]
```diff
@@ -1,294 +1,394 @@
...
 from simtools.ray_tracing.ray_tracing import RayTracing
+from simtools.utils import names
+from simtools.visualization import plot_psf
+
+
+@dataclass
+class MirrorOptimizationResult:
+    """Dataclass to store the result of a single mirror RNDA optimization."""
+
+    mirror: int  # One-based mirror index
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/mirror_panel_psf.py]: Currently the code isnt reading mirror_id (the data we got in ```198mir_190925.ecsv``` is just a table with 198 entries for d80) so I'm just assuming the index of the psf is the mirror id. Now changes it to zero-based as well.
